### PR TITLE
Feature: VSDSPUB-614 index on Member

### PR DESCRIPTION
**Pull request by WLefever-Cegeka**
Adds an index on members. Index goes for every collection from 1 to ...

### CODE REVIEW

**Review by Yalz — COMMENTED** [review]
Having this sequence seems really bloated, can't we use a simple atomicinteger for this?

File: ldes-fragmentisers/ldes-fragmentisers-geospatial/src/test/java/be/vlaanderen/informatievlaanderen/ldes/server/fragmentisers/geospatial/bucketising/GeospatialBucketiserTest.java [hunk 1]
```diff
@@ -68,9 +68,9 @@ private Member readLdesMemberFromFile(ClassLoader classLoader, String fileName)
 				.fromString(Files.lines(Paths.get(file.toURI())).collect(Collectors.joining())).lang(Lang.NQUADS)
 				.toModel();
 
-		return new Member("collectionName",
-				"https://private-api.gipod.beta-vlaanderen.be/api/v1/mobility-hindrances/10810464/1",
-				null, null, outputModel, List.of());
+		return new Member("https://private-api.gipod.beta-vlaanderen.be/api/v1/mobility-hindrances/10810464/1",
+				"collectionName",
```
[the remark this answers is not in the document]
Comment (WLefever-Cegeka) [reply ldes-fragmentisers/ldes-fragmentisers-geospatial/src/test/java/be/vlaanderen/informatievlaanderen/ldes/server/fragmentisers/geospatial/bucketising/GeospatialBucketiserTest.java:72]: Personally, it seems more logical to have Id as the first attribute of Member and other attributes such as collectionName afterwards.

**Comment by WLefever-Cegeka** [comment]
> omicinteger for t

https://www.baeldung.com/spring-boot-mongodb-auto-generated-field

**Comment by WLefever-Cegeka** [comment]
> Having this sequence seems really bloated, can't we use a simple atomicinteger for this?

This seems like a standardized approach:
* https://www.baeldung.com/spring-boot-mongodb-auto-generated-field
* https://bootify.io/mongodb/generating-ids-spring-data-mongodb-with-sequences.html

File: ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/tree/member/entities/Member.java [hunk 2]
```diff
@@ -10,17 +10,19 @@
...
 	private final String versionOf;
 	private final LocalDateTime timestamp;
 	private final List<String> treeNodeReferences;
 
-	public Member(String collectionName, String memberId, String versionOf, LocalDateTime timestamp,
+	public Member(String memberId, String collectionName, Long index, String versionOf, LocalDateTime timestamp,
 			final Model memberModel, List<String> treeNodeReferences) {
 		this.collectionName = collectionName;
 		this.memberId = memberId;
+		this.index = index;
```
Comment (Tomvbe) [inline ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/tree/member/entities/Member.java:25]: I don't know if 'index' is the best name for this. I associate this more with sequences than indices. Wouldn't sequenceNr  be more explicite?

File: ldes-server-infra-mongo/mongo-repository/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/infra/mongo/member/service/LdesMemberEntityListener.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package be.vlaanderen.informatievlaanderen.ldes.server.infra.mongo.member.service;
+
+import be.vlaanderen.informatievlaanderen.ldes.server.infra.mongo.member.entity.LdesMemberEntity;
+import be.vlaanderen.informatievlaanderen.ldes.server.infra.mongo.membersequence.service.SequenceGeneratorService;
+import org.springframework.data.mongodb.core.mapping.event.AbstractMongoEventListener;
+import org.springframework.data.mongodb.core.mapping.event.BeforeConvertEvent;
+import org.springframework.stereotype.Component;
+
+@Component
+public class LdesMemberEntityListener extends AbstractMongoEventListener<LdesMemberEntity> {
```
Comment (Tomvbe) [inline ldes-server-infra-mongo/mongo-repository/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/infra/mongo/member/service/LdesMemberEntityListener.java:10]: This name is quite generic for a very specific listener, however I can't really come up with a better name without making it super long like LdesMemberEntityMongoEventListener :(

File: ldes-server-infra-mongo/mongo-repository/src/test/java/be/vlaanderen/informatievlaanderen/ldes/server/infra/mongo/membersequence/service/SequenceGeneratorServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.springframework.data.mongodb.core.query.Criteria.where;
+import static org.springframework.data.mongodb.core.query.Query.query;
+
+class SequenceGeneratorServiceTest {
```
Comment (Tomvbe) [inline ldes-server-infra-mongo/mongo-repository/src/test/java/be/vlaanderen/informatievlaanderen/ldes/server/infra/mongo/membersequence/service/SequenceGeneratorServiceTest.java:16]: maybe we can use a test to make sure that two collections have separate sequences?

**Review by Tomvbe — APPROVED** [review]

File: ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/tree/member/entities/Member.java (hunk 2 above)
Comment (WLefever-Cegeka) [reply ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/tree/member/entities/Member.java:25]: Updated to `sequenceNr`

File: ldes-server-infra-mongo/mongo-repository/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/infra/mongo/member/service/LdesMemberEntityListener.java (hunk 3 above)
Comment (WLefever-Cegeka) [reply ldes-server-infra-mongo/mongo-repository/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/infra/mongo/member/service/LdesMemberEntityListener.java:10]: quote is on variables, but I guess it also applies to classes :-)

“You should name a variable using the same care with which you name a first-born child.”

― Robert C. Martin, [Clean Code: A Handbook of Agile Software Craftsmanship]

File: ldes-server-infra-mongo/mongo-repository/src/test/java/be/vlaanderen/informatievlaanderen/ldes/server/infra/mongo/membersequence/service/SequenceGeneratorServiceTest.java (hunk 4 above)
Comment (WLefever-Cegeka) [reply ldes-server-infra-mongo/mongo-repository/src/test/java/be/vlaanderen/informatievlaanderen/ldes/server/infra/mongo/membersequence/service/SequenceGeneratorServiceTest.java:16]: We should have an integration test for this, but I'm noticed that the Mongo integration tests are not working:
https://github.com/Informatievlaanderen/VSDS-LDESServer4J/issues/510

I updated the issue with a request to also add integration tests for this.

**Pull request merged**